containerd: resource-limit settings for oci default (#3206)
bottlerocket-os/bottlerocket: README.md

The full list of resource limits that can be configured in Bottlerocket are:

resource limit | setting | default value | unit
----- | ----- | ----- | -----
`RLIMIT_AS` | `settings.oci-defaults.resource-limits.max-address-space.hard-limit` | - | bytes

> `RLIMIT_AS` | `settings.oci-defaults.resource-limits.max-address-space.soft-limit` | - | bytes

Reviewer: If we use an HTML table instead, we could use `rowspan=2` for the resource limit and unit fields to avoid duplicating those values. Might be worth trying out to see if the result seems more polished.

`RLIMIT_MSGQUEUE` | `settings.oci-defaults.resource-limits.max-msgqueue-size.hard-limit` | - | bytes
`RLIMIT_MSGQUEUE` | `settings.oci-defaults.resource-limits.max-msgqueue-size.soft-limit` | - | bytes
`RLIMIT_NICE` | `settings.oci-defaults.resource-limits.max-nice-priority.hard-limit` | - | -
`RLIMIT_NICE` | `settings.oci-defaults.resource-limits.max-nice-priority.soft-limit` | - | -
`RLIMIT_NOFILE` | `settings.oci-defaults.resource-limits.max-open-files.hard-limit` | 1048576 | files
`RLIMIT_NOFILE` | `settings.oci-defaults.resource-limits.max-open-files.soft-limit` | 65536 | files
`RLIMIT_SIGPENDING` | `settings.oci-defaults.resource-limits.max-pending-signals.hard-limit` | - | signals
`RLIMIT_SIGPENDING` | `settings.oci-defaults.resource-limits.max-pending-signals.soft-limit` | - | signals
`RLIMIT_NPROC` | `settings.oci-defaults.resource-limits.max-processes.hard-limit` | - | processes
`RLIMIT_NPROC` | `settings.oci-defaults.resource-limits.max-processes.soft-limit` | - | processes
`RLIMIT_RTPRIO` | `settings.oci-defaults.resource-limits.max-realtime-priority.hard-limit` | - | -
`RLIMIT_RTPRIO` | `settings.oci-defaults.resource-limits.max-realtime-priority.soft-limit` | - | -

> `RLIMIT_RTTIME` | `settings.oci-defaults.resource-limits.max-realtime-timeout.hard-limit` | - | us
> `RLIMIT_RTTIME` | `settings.oci-defaults.resource-limits.max-realtime-timeout.soft-limit` | - | us

Reviewer: nit: `microseconds` might be more readable than `us`:
```suggestion
`RLIMIT_RTTIME` | `settings.oci-defaults.resource-limits.max-realtime-timeout.hard-limit` | - | microseconds
`RLIMIT_RTTIME` | `settings.oci-defaults.resource-limits.max-realtime-timeout.soft-limit` | - | microseconds
```

Please see the [`getrlimit` linux manpage](https://man7.org/linux/man-pages/man7/capabilities.7.html) for meanings of `hard-limit` and `soft-limit`.

The full list of resource limits that can be configured in Bottlerocket are:

> <table>

Reviewer 2: It's highly unlikely we will want to render into anything other than HTML, but is there a reason for using HTML tags directly rather than sticking with markdown table syntax?

Author: So that we can use rowspan. This will avoid the repetition of `RLIMIT_NOFILE`  rlimit names.

<tr>

>   <th>resource limit</th>

Reviewer 3: nit: capiltalize for each first letter

  <th>setting</th>
  <th>default value</th>
  <th>unit</th>
</tr>
<tr>
  <td rowspan="2">RLIMIT_AS</td>

>   <td>settings.oci-defaults.resource-limits.max-address-space.hard-limit</td>

Reviewer 3: You are missing " \`\` "  in all the settings. Same for `RLIMIT_<>`.

  <td>-</td>

>   <td>bytes</td>

Reviewer: ```suggestion
  <td rowspan="2">bytes</td>
```

  <td><code>settings.oci-defaults.resource-limits.max-realtime-priority.hard-limit</code></td>
  <td>-</td>
  <td rowspan="2">-</td>
</tr>
<tr>
  <td><code>settings.oci-defaults.resource-limits.max-realtime-priority.soft-limit</code></td>
  <td>-</td>
</tr>
<tr>
  <td rowspan="2"><code>RLIMIT_RTTIME</code></td>
  <td><code>settings.oci-defaults.resource-limits.max-realtime-timeout.hard-limit</code></td>
  <td>-</td>
  <td rowspan="2">microseconds</td>
</tr>
<tr>
  <td><code>settings.oci-defaults.resource-limits.max-realtime-timeout.soft-limit</code></td>
  <td>-</td>
</tr>
<tr>
  <td rowspan="2"><code>RLIMIT_RSS</code></td>
  <td><code>settings.oci-defaults.resource-limits.max-resident-set.hard-limit</code></td>
  <td>-</td>
  <td rowspan="2">bytes</td>
</tr>
<tr>
  <td><code>settings.oci-defaults.resource-limits.max-resident-set.soft-limit</code></td>
  <td>-</td>
</tr>
<tr>
  <td rowspan="2"><code>RLIMIT_STACK</code></td>
  <td><code>settings.oci-defaults.resource-limits.max-stack-size-limit</code></td>
  <td>-</td>
  <td rowspan="2">bytes</td>
</tr>
<tr>
  <td><code>settings.oci-defaults.resource-limits.max-stack-size-limit</code></td>
  <td>-</td>
</tr>
</table>

> **Note**:  String `unlimited` can be used to set resource-limit as unlimited as

Reviewer 4: ```suggestion
**Note**:  The string `unlimited` can be used to set resource-limit as unlimited. For example, to set the hard limit to unlimited:
```

  <td><code>settings.oci-defaults.resource-limits.max-locked-memory.hard-limit</code></td>
  <td>-</td>
  <td rowspan="2">bytes</td>
</tr>
<tr>
  <td><code>settings.oci-defaults.resource-limits.max-locked-memory.soft-limit</code></td>
  <td>-</td>
</tr>
<tr>
  <td rowspan="2"><code>RLIMIT_MSGQUEUE</code></td>
  <td><code>settings.oci-defaults.resource-limits.max-msgqueue-size.hard-limit</code></td>
  <td>-</td>
  <td rowspan="2">bytes</td>
</tr>
<tr>
  <td><code>settings.oci-defaults.resource-limits.max-msgqueue-size.soft-limit</code></td>
  <td>-</td>
</tr>
<tr>
  <td rowspan="2"><code>RLIMIT_NICE</code></td>
  <td><code>settings.oci-defaults.resource-limits.max-nice-priority.hard-limit</code></td>
  <td>-</td>
  <td rowspan="2">-</td>
</tr>
<tr>
  <td><code>settings.oci-defaults.resource-limits.max-nice-priority.soft-limit</code></td>
  <td>-</td>
</tr>
<tr>
  <td rowspan="2"><code>RLIMIT_NOFILE</code></td>
  <td><code>settings.oci-defaults.resource-limits.max-open-files.hard-limit</code></td>
  <td>1048576</td>
  <td rowspan="2">files</td>
</tr>
<tr>
  <td><code>settings.oci-defaults.resource-limits.max-open-files.soft-limit</code></td>
  <td>65536</td>

> </tr>

Reviewer: nit: having "soft" come before "hard" would sort the numbers numerically, but should be done for all other limits too.

  <td><code>settings.oci-defaults.resource-limits.max-realtime-priority.hard-limit</code></td>
  <td>-</td>
  <td rowspan="2">-</td>
</tr>
<tr>
  <td><code>settings.oci-defaults.resource-limits.max-realtime-priority.soft-limit</code></td>
  <td>-</td>
</tr>
<tr>
  <td rowspan="2"><code>RLIMIT_RTTIME</code></td>
  <td><code>settings.oci-defaults.resource-limits.max-realtime-timeout.hard-limit</code></td>
  <td>-</td>
  <td rowspan="2">microseconds</td>
</tr>
<tr>
  <td><code>settings.oci-defaults.resource-limits.max-realtime-timeout.soft-limit</code></td>
  <td>-</td>
</tr>
<tr>
  <td rowspan="2"><code>RLIMIT_RSS</code></td>
  <td><code>settings.oci-defaults.resource-limits.max-resident-set.hard-limit</code></td>
  <td>-</td>
  <td rowspan="2">bytes</td>
</tr>
<tr>
  <td><code>settings.oci-defaults.resource-limits.max-resident-set.soft-limit</code></td>
  <td>-</td>
</tr>
<tr>
  <td rowspan="2"><code>RLIMIT_STACK</code></td>
  <td><code>settings.oci-defaults.resource-limits.max-stack-size-limit</code></td>
  <td>-</td>
  <td rowspan="2">bytes</td>
</tr>
<tr>
  <td><code>settings.oci-defaults.resource-limits.max-stack-size-limit</code></td>
  <td>-</td>

> </tr>

Reviewer: ```suggestion
<tr>
  <td rowspan="2"><code>RLIMIT_STACK</code></td>
  <td><code>settings.oci-defaults.resource-limits.max-stack-size.hard-limit</code></td>
  <td>-</td>
  <td rowspan="2">bytes</td>
</tr>
<tr>
  <td><code>settings.oci-defaults.resource-limits.max-stack-size.soft-limit</code></td>
  <td>-</td>
</tr>
```

</table>

> Any integer between 0 to 4294967295 (2 ^32^ -1) and `unlimited` are valid values to set hard-limit and soft limit for a resource. 

Reviewer 3: nit: missing `-` in "soft limit"

The string `unlimited` as resource-limit denotes limit as unlimited. Following are the examples to set hard-limit and soft-limit:
* Setting any valid integer value as limit:
  ```toml
  [settings.oci-defaults.resource-limits.<rlimit>>]
  hard-limit = 65536
  soft-limit = 65536

>     ```

Reviewer 3: nit: There is an extra space here, and Github paints all of this red.

</tr>
<tr>
  <td><code>settings.oci-defaults.resource-limits.max-realtime-priority.hard-limit</code></td>
  <td>-</td>
</tr>
<tr>
  <td rowspan="2"><code>RLIMIT_RTTIME</code></td>
  <td><code>settings.oci-defaults.resource-limits.max-realtime-timeout.soft-limit</code></td>
  <td>-</td>
  <td rowspan="2">microseconds</td>
</tr>
<tr>
  <td><code>settings.oci-defaults.resource-limits.max-realtime-timeout.hard-limit</code></td>
  <td>-</td>
</tr>
<tr>
  <td rowspan="2"><code>RLIMIT_RSS</code></td>
  <td><code>settings.oci-defaults.resource-limits.max-resident-set.soft-limit</code></td>
  <td>-</td>
  <td rowspan="2">bytes</td>
</tr>
<tr>
  <td><code>settings.oci-defaults.resource-limits.max-resident-set.hard-limit</code></td>
  <td>-</td>
</tr>
<tr>
  <td rowspan="2"><code>RLIMIT_STACK</code></td>
  <td><code>settings.oci-defaults.resource-limits.max-stack-size.soft-limit</code></td>
  <td>-</td>
  <td rowspan="2">bytes</td>
</tr>
<tr>
  <td><code>settings.oci-defaults.resource-limits.max-stack-size.hard-limit</code></td>
  <td>-</td>
</tr>
</table>

Limits can be any integer between 0 to `int64::MAX`. Either `-1` or `"unlimited"` can be used to remove the limit.

> * Specifying the maximum value (`u64::MAX`) for a limit:

Reviewer: ```suggestion
* Specifying the maximum value (`i64::MAX`) for a limit:
```

  <td><code>settings.oci-defaults.resource-limits.max-realtime-priority.soft-limit</code></td>
  <td>-</td>
  <td rowspan="2">-</td>
</tr>
<tr>
  <td><code>settings.oci-defaults.resource-limits.max-realtime-priority.hard-limit</code></td>
  <td>-</td>
</tr>
<tr>
  <td rowspan="2"><code>RLIMIT_RTTIME</code></td>
  <td><code>settings.oci-defaults.resource-limits.max-realtime-timeout.soft-limit</code></td>
  <td>-</td>
  <td rowspan="2">microseconds</td>
</tr>
<tr>
  <td><code>settings.oci-defaults.resource-limits.max-realtime-timeout.hard-limit</code></td>
  <td>-</td>
</tr>
<tr>
  <td rowspan="2"><code>RLIMIT_RSS</code></td>
  <td><code>settings.oci-defaults.resource-limits.max-resident-set.soft-limit</code></td>
  <td>-</td>
  <td rowspan="2">bytes</td>
</tr>
<tr>
  <td><code>settings.oci-defaults.resource-limits.max-resident-set.hard-limit</code></td>
  <td>-</td>
</tr>
<tr>
  <td rowspan="2"><code>RLIMIT_STACK</code></td>
  <td><code>settings.oci-defaults.resource-limits.max-stack-size.soft-limit</code></td>
  <td>-</td>
  <td rowspan="2">bytes</td>
</tr>
<tr>
  <td><code>settings.oci-defaults.resource-limits.max-stack-size.hard-limit</code></td>
  <td>-</td>
</tr>
</table>

> Any integer between 0 to 9223372036854775807 (2 ^63^ -1), -1 and `unlimited` are valid values to set hard-limit and soft limit for a resource. 

Reviewer: ```suggestion
Limits can be any integer between 0 to `int64::MAX`. Either `-1` or `"unlimited"` can be used to remove the limit.
```

> The string `unlimited` or -1 as resource-limit denotes limit as unlimited. Following are the examples to set hard-limit and soft-limit:

Reviewer: ```suggestion
```

> * Setting any valid integer value as limit:
>   ```toml
>   [settings.oci-defaults.resource-limits.<rlimit>>]
>   hard-limit = 65536
>   soft-limit = 65536
>   ```

Reviewer: This is implied by the next example.
```suggestion
```

> * Setting max possible integer value as limit:
>   ```toml
>   [settings.oci-defaults.resource-limits.<rlimit>>]
>   hard-limit = 9223372036854775807
>   soft-limit = 65536
>   ```

Reviewer: ```suggestion
* Specifying the maximum value (`u64::MAX`) for a limit:
  ```toml
  [settings.oci-defaults.resource-limits.<rlimit>]
  soft-limit = 65536
  hard-limit = 9223372036854775807
  ```
```

> * Setting unlimited value as limit:
>   ```toml
>   [settings.oci-defaults.resource-limits.<rlimit>>]
>   hard-limit = "unlimited"
>   soft-limit = 65536

Reviewer: ```suggestion
* Removing a limit:
  ```toml
  [settings.oci-defaults.resource-limits.<rlimit>]
  soft-limit = 65536
  hard-limit = "unlimited"
```

  ```
* Setting unlimited value as limit:
  ```toml
  [settings.oci-defaults.resource-limits.<rlimit>>]
  hard-limit = -1
  soft-limit = 65536

>   ```

Reviewer: We can support this in code, but don't need to document it. The "unlimited" one is more obvious and should be preferred.
